\documentclass[a4paper,12pt]{article}
\usepackage[english]{babel}
\usepackage{graphicx}
\usepackage{csvsimple}
\usepackage{pgfplotstable}
\usepackage{booktabs}
\usepackage{chngcntr}
\usepackage{sectsty}
\usepackage{pdfpages}
\usepackage{longtable}
\usepackage{tocloft}
\usepackage{amsmath}

%\usepackage[pdftex]{graphicx}
\usepackage[ansinew]{inputenc}
\usepackage{geometry}
%\usepackage{bbold}
\geometry{a4paper,left=2.5cm,right=2.5cm, top=2.5cm, bottom=3cm}
\newcommand{\HRule}{\rule{\linewidth}{0.5mm}}

\usepackage{textcomp}
\usepackage[hyphens,spaces]{url}

\begin{document}
\sloppy
\sectionfont{\LARGE}
\subsectionfont{\Large}
\subsubsectionfont{\Large}
\counterwithin{figure}{section}
\counterwithin{table}{section}

\begin{titlepage}


%%LR
\sffamily

\begin{center}


% Oberer Teil der Titelseite:
\includegraphics[width=0.3\textwidth]{../figures/LMU-logo.jpg}
\hfill
\includegraphics[width=0.4\textwidth]{../figures/TUM-logo.jpg}  
\\[5cm]

{\LARGE Department of Bioinformatics and Computational Biology}\\[0.5cm]
{Technische Universit\"at M\"unchen}\\
[2cm]
{\Large Master's Thesis in Bioinformatics}\\[1.5cm]

% Title
\HRule \\[0.4cm]
{ \huge \bfseries Variation of HERV elements in the KORA cohort
}\\[0.4cm]

\HRule \\[1.5cm]

{\Large Julian Schmidt}\\[2.5cm]

\vfill
\end{center}
\end{titlepage}
\pagestyle{empty}

%%LR comprehensive title
\begin{titlepage}
{\sffamily


\begin{center}
\includegraphics[width=0.3\textwidth]{../figures/LMU-logo.jpg}
\hfill
\includegraphics[width=0.4\textwidth]{../figures/TUM-logo.jpg}  
\\[1.5cm]  

{\LARGE Department of Bioinformatics and Computational Biology}\\[0.5cm]
{Technische Universit\"at M\"unchen}\\[1cm]

{\Large Master's Thesis in Bioinformatics}\\[2cm]
{\textbf{\LARGE Variation of HERV elements in the KORA cohort}}\\[2cm]
{\textbf{\LARGE Variation von HERV elementen in der KORA Kohorte}}\\[4cm]

\end{center}
\begin{center}\Large
  \begin{tabular}{ll}
    Author:& Julian Schmidt\\
    Supervisor: & Dr. Matthias Heinig\\
    Advisor:        & Johann Hawe\\
    Submitted:     &  15.03.2018\\
  \end{tabular}
\end{center}

}% end title page

\end{titlepage}

\pagenumbering{gobble}

\tableofcontents
\newpage

\pagestyle{plain}
\pagenumbering{arabic}
\section{Introduction}

\subsection{Regulation of cell functions and Epigenetics}
\begin{itemize}
\item central dogma of biology
\item rising importance of other factors atop DNA \textrightarrow epigenetics
\item quick overview epigenetic marks
\item chromatin states
\item DNA methylation 
\item snp -> cpg -> expression pattern
\item TF <-> cpg interaction
\end{itemize}


\subsection{Human endogenous retroviruses}
\begin{itemize}
\item first humane genome -> "junk DNA"
\item ongoing discovery for non-coding regions
\item still masking of difficult sequence for many analysis -> repeats
\item repeat classes -> ... -> herv
\item herv origin - ...-virus like 
\item herv structure: LTR - pol - env - ... - LTR
\item discovered roles of hervs in general regulation/diseases
\item 
\end{itemize}

\subsection{Effect network analysis}
\begin{itemize}
\item many bioinformatics methods find correlations, but not direct cause
\item attempt to discern direct connections from bigger data webs
\item hope to find possible biological mechanisms of gene regulation = path in model 
\item used approach: Gaussian Graphical models
\end{itemize}


\newpage
\section{Data}


\subsection{HERV annotation}
HERV annotation was pertained from RepeatMasker\cite{RepeatMasker} repeat library. RepeatMasker is a tool that screens DNA sequences against a library interspersed repeats and low complexity DNA sequences. It generates an annotation of identified repeats and masks them in the query sequence.

The track representing all identified elements from the RepeatMasker library for human genome hg19 was downloaded from the UCSC genome browser download section \url{(http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/rmsk.txt.gz)}. It contains a total of 5298130 occurrences of repeats. Each entry consists of 17 values. Not in order these are repeat name the repeat class and family, as well as the chromosome, strand, and the genomic start and end position of the repeat occurrence. Furthermore it contains the position in the whole sequence, that is known for the identified repeat, that is covered by the occurrence. It also describes the quality of the alignment of the repeat sequence to the annotated position using the Smith Waterman alignment score\cite{SMITH1981195} and the number of base mismatches, deletions and insertions per thousand base pairs. Finally there is an indexing field used to speed up chromosome range queries and the first digit of the id field in the RepeatMasker output file. The first 10 lines of the track are shown in table \ref{tab:RepeatMaskerTrack}.

\begin{table}[b]
\centering
\label{tab:RepeatMaskerTrack}
\resizebox{\textwidth}{!}{
\pgfplotstabletypeset[
    col sep=tab, 
    header=true,    
	columns={bin, swScore, milliDiv, milliDel, milliIns, genoName, genoStart, genoEnd, genoLeft, strand, repName, repClass, repFamily, repStart, repEnd, repLeft, id},
	columns/genoName/.style={column type=l,string type},
	columns/strand/.style={column type=l,string type},
	columns/repName/.style={column type=l,string type, string replace*={_}{\textunderscore}},
	columns/repClass/.style={column type=l,string type, string replace*={_}{\textunderscore}},
	columns/repFamily/.style={column type=l,string type, string replace*={_}{\textunderscore}},
	every head row/.style={before row=\hline, after row=\hline},
	every last row/.style={after row=\hline}]{../tables/head.repeatmasker.track.tsv}}
\caption{First ten rows of the RepeatMasker annotation on hg19}
\end{table}

To extract HERV elements the annotation was filtered on different columns generating three sets variable size. Multiple HERV sets were created as an attempt to cover different possible definitions of HERV elements.

This work only discerned between different HERV element types by defining different HERV sets. Therefore, within each set annotations, whose genomic positions overlap or are directly adjacent, were merged into one element.

A first set, HERV set 1 (HERV S1) was constructed by extracting all elements that contained "ERV" in the repeat name column. This set
The resulting 42508 annotations condensed to 35589 elements after merging. The elements have a mean width of 949 bp and cover a total of 33.8 Mbp, which is ca 1.04\% of the human genome. The distribution of element lengths in HERV S1 is shown in Figure \ref{fig:hervS1.lengths.hist}  

Alternatively filtering the annotation for "ERV" in the super family column leads to 696689 annotations. HERV set 2 (HERV S2) contains all endogenous retroviral sequences found in the human genome. After merging overlapping and adjacent annotations this led to 633323 elements. Their mean width is 415 bp and they make up to 262.8 Mbp or ca 8.13\% of the human genome. 

A third set "HERV set 3" (HERV S3) was constructed by filtering the repeat name column for "HERV", which resulted in 21361 annotations. As this set contains only elements that are explicitly named "HERV", it is my hope that might contain only endogenous retroviral elements, that were inserted directly into the human genome. Merging overlapping and adjacent annotations resulted in 15403 elements with an average width of 1468 bp and a combined length of 22.6 Mbp.

\begin{figure}[tb]
	\includegraphics[scale = 1, keepaspectratio = true]{../figures/hervS1_lengths_hist}  
	\caption{Length distribution of elements in HERV S1}
    \label{fig:hervS1.lengths.hist}
\end{figure}


\subsection{KORA}
The expression, methylation and genotype data used in this work were generated by the platform for Cooperative Health Research in the Region of Augsburg - short KORA. It contains health surveys as well as examinations of individuals of German nationality living in the area of Augsburg, Bavaria.
The objective of KORA is to track changes in health conditions over a long period in order to identify and examine the causes, effects and development of chronical diseases.

The data comes from the KORA F4 Survey, which was conducted from 2006 to 2008 and comprised samples of 3080 individuals. F4 is a follow up study to the survey S4 performed from 1999 to 2001 and containing 4261 individuals. 

All measurements were performed on whole blood samples. Houseman blood counts\cite{} describing the composition of different cell types for each individual are available.

Not all essays are available for all samples. Therefore different analyses were performed on varying sets of individuals according to availability of the required data types. A diagram of the which samples were available for each essay can be seen in figure \ref{fig:samples.venn}.

\begin{figure}[tb]
	\includegraphics[scale = 1, keepaspectratio = true]{../figures/samples_venn}  
	\caption{Number of samples with genotype, expression and methylation measurements in KORA F4 Survey}
    \label{fig:samples.venn}
\end{figure}

\subsubsection{Expression}
The expression data was generated using the HumanHT-12 v3.0 Gene Expression BeadChip. The chip can measure expression values for 49576 probes. However only 47864 probes represent an actual genomic location. 

Measurements for 993 individuals are available from the KORA F4 survey. The comprise values for a total of 48803 probes per sample. Probes that do not map to a genomic location were excluded in all analyses, leaving 47864 probes. Of these 29521 are annotated to total of 19288 genes. 

To not lose information, especially in hERV regions that are usually sparsely annotated with genes, probes without genes were not discarded and most analyses were performed on probe level or only partially abstracting to gene level. 

\begin{figure}[tb]
	\includegraphics[scale = 1, keepaspectratio = true]{../figures/expr_var}  
	\caption{Coefficient of expression variance over 993 individuals}
    \label{fig:expr.var}
\end{figure}

\subsubsection{Methylation}
DNA methylation was measured using the Infinium HumanMethylation450K BeadChip, which interrogates methylation levels at 485577 genomic locations.  

Methylation data was available for 1727 individuals and 485512 sites, which make up all 'cg' and 'ch' probe type probes.
 
\subsubsection{Genotypes}
Genotyping was performed with the Affymetrix Axiom array. The Iluminus calling algorithm was used for genotype calling and missing values were imputed using the IMPUTE2 software\cite{10.1371/journal.pgen.1000529}. SNPs were filtered at IMPUTE value of 0.4. 

After excluding all SNPs with a minor allele frequency of less than one percent measurements of 9533127 SNPs for 3788 individuals were available. 

\subsubsection{Covariates}
Several covariates were known for each sample. These were age, sex, body mass index (BMI) and wide blood cell count, as well as experimental factors like storage time and RNA integrity number (RIN).

\subsubsection{Methylation quantitative trait loci}
Previously process methylation quantitative trait loci (meQTL) data was used. The data set contained a total of xxxxxx significantly associated cpg-snp pairs. 
xxxx distinct cpg-sites and xxxxxx snps were part of at least one meQTL. xxxx pairs consisted of cpgs and snps on the same chromosome, while xxxxx association were between different chromosomes. 

\subsection{Transcription factor binding}
Transcription factor binding sites were obtained from two publicly available sources: 

First was the third version of the track "Transcription Factor ChIP-seq (161 factors) from ENCODE with Factorbook Motifs"\cite{10.1101/gr.139105.112} downloaded from the UCSC genome browser download section (\url{http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeRegTfbsClustered/wgEncodeRegTfbsClusteredWithCellsV3.bed.gz}).

It combines 690 high quality ENCODE ChIP-seq data sets, which were processed with the Factorbook motif discovery and annotation pipeline\cite{10.1101/gr.139105.112}. The pipeline uses the tools MEME-ChIP\cite{10.1093/bioinformatics/btr189} and FIMO\cite{10.1093/bioinformatics/btr064} from the MEME software suite and merges discovered motifs with known motifs from Jaspar\cite{10.1093/nar/gkx1126} and TransFac\cite{10.1093/nar/gkj143} using machine learning methods and manual curation. 

The track contains a total of 438044 distinct peaks for 161 transcription factors in 91 cell types. For our analyses we filtered and combined the peaks for 23 blood related cell types. This leaves a total of 2173371 peaks for 125 transcription factors.  

The second source was the ReMap project\cite{10.1093/nar/gku1280}. It combines 395 publicly available ChIP-seq data sets covering 132 different transcription factors across 83 cell lines. ReMap uses Bowtie2\cite{} to map reads to the human genome and the tool MACS\cite{} for peak calling. The finished data set was downloaded from the ReMap website\url{(http://tagc.univ-mrs.fr/remap/download/All/filPeaks_public.bed.gz)}.

It contains xxxxx peaks. After filtering for 19 blood related cell types a of xxxx peaks of xxx different transcription factors remained. 

Combining both filtered data sets lead to a total of xxxxx peaks of xxx transcription factors.


\subsection{Chromatin states}
Chromatin state annotations were downloaded from Roadmap Epgenomics Core 15-state model\url{http://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final}. The Model provides a whole genome chromatin state annotation of 200 bp wide windows to the following 15 states: Active Transcription Start Site (TSS), Flanking Active TSS, Transcription at gene 5' and 4', Strong transcription, Weak transcription, Genic enhancers, Enhancers, ZNF genes \& repeats, Heterochromatin, Bivalent/Poised TSS, Flanking Bivalent TSS/Enhancer, Bivalent Enhancer, Repressed PolyComb, Weak Repressed PolyComb and Quiescent/Low. The model is available for 127 diverse cell lines. 

It was generated using ChromHMM v1.10\cite{10.1038/nmeth.1906} on the chromatin marks H3K4me1, H3K4me3, H3K27me3, H3K9me3, and H3K36me3. ChromHMM is based on a multivarate Hidden Markov Model. 

In this work the annotations for 27 blood related cell lines were used. To get the distribution of states for a single feature or a set of features the different annotations for the different cell lines were weighted according to the houseman counts and summed up.

\newpage
\section{Methods}
\subsection{Overlaps}
The task of whether HERV elements contained any annotated elements like expression probes, genes, cpg sites and transcription factor binding sites (tfbs) was performed using function "findOverlaps" from the Bioconductor package "GenomicRanges"\cite{10.1371/journal.pcbi.1003118}. 

Features were considered to be of interest for the analysis of an HERV element, if they overlapped by at least one base pair.

Features were also filtered for the HERV elements and their 1kb or 2kb up- and downstream regions.

\subsection{Data normalization}
Expression and methylation values were corrected for available covariates by calculating the residual matrix. 
A linear model containing the cell compositions and the first 20 principal components was used for methylation.

Expression residuals were calculated using a linear model of age, sex, RNA integrity number (RIN), plate and storage time. 

\subsection{eQTL/eQTM calculation}
Expression quantitative trait loci (eQTL) and expression quantitative trait methylation were calculated using the Bioconductor package MatrixEQTL\cite{10.1093/bioinformatics/bts163}. MatrixEQTL tests for association of SNP-transcript pairs. It offers two modes of modeling the effect of the genotype on transcription levels: 

When setting the parameter "$useModel = modelLINEAR$", as was done in this work, an additive linear model is used. The association ism modeled as simple linear regression and the absolute value of the sample correlation is used as test statistic.
 
Alternatively when choosing "$modelAnova$" for the parameter, the effect is modeled with ANOVA model. In this case the test statistic is the squared sample correlation.

After calculating the test statistics the p-values for the all pairs that pass a defined significance threshold are calculated. These are corrected multiple testing using a Benjamini- Hochberg procedure, adapted for not recording all p-values.

MatrixEQTL is very performant because it manages to reduce the calculation of the test statistic to one single large matrix multiplication by cleverly transforming the genotype and transcription variables.

MatrixEQTL also allows to include covariates in the QTL calculation. As the expression and methylation values used are residuals and therefore already consider covariates this option is not used. Furthermore MatrixEQTL can differentiate between cis- and trans-interactions. The maximal distance to consider a pair on the same chromosome as cis was set to 50kb. 

The threshold for significant cis-QTLs during calculation was set to 10e-6 and 10e-8 for trans.

\subsection{Functional Analysis of Gene Sets}
In multiple analyses functional Gene Ontology enrichments were performed.

First a set of all GO annotations with any evidence code for gene symbols  was retrieved from the Bioconductor package AnnotationDbi\cite{AnnotationDbi}. 

Then a hypergeometric test\cite{GOstats} for overreprensentation is performed on a set of genes of interest. For most enrichments a custom background set of genes specific to the analysis is given. Finally the p-values for overrepresented GO terms were adjusted for multiple testing using the Holm method\cite{10.2307/4615733}. Terms with an adjusted p-value of less than 0.05 were considered significantly overrepresented. 

\subsection{Gaussian Graphical Models}




\newpage
\section{Results}
\subsection{Normalized Data}
The distribution of the quantile normalized expression values and the expression residuals over all probes and samples can be seen in figure \ref{fig:expr.raw.res}. As expected the residuals follow a normal distribution. This is important for the calculation of the Gaussian graphical models, as the normal distribution of the data is one of base assumptions\cite{}.

\subsection{HERV region features}
In this section I will describe the expression probes that overlap with any HERV element and the cpgs and SNPs that lie within any HERV element and/or their flanking regions. The results for the set of all endogenous retroviral elements, HERV set 2, without flanking regions are described in detail. The results for the other sets and including flanking regions will be shown in tables or in the supplementary data.

\subsubsection{Expression}
A total of 2343 expression probes overlap directly with at least one of 2271 HERV elements from HERV S2. This means ca 4.50\% of all expression probes overlap with HERV elements and expression measurements are known for parts of ca 0.37\% of all HERV elements.

The mean expression of the probes overlapping with HERV elements, as shown in figure \ref{fig:hervS2.expr.var} is on average lower than for all probes (figure \ref{fig:expr.var}). The coefficient of variance, however, shows a very similar pattern for the whole set of probes and the considered subset.

\begin{figure}[tb]
	\includegraphics[scale = 1, keepaspectratio = true]{../figures/hervS2_expr_var}  
	\caption{Coefficient of expression variance over 993 individuals and 2343 expression probes overlapping with HERV S2.}
    \label{fig:hervS2.expr.var}
\end{figure}

Of these 2343 probes 510 were annotated to one of 449 different genes. I performed a GO enrichment for the biological process ontology on these genes with the set of all genes with available expression data as background. After correcting for multiple testing only the term defense response (GO:0006952, p-value=1.37e-5, fdr=0.047) was significantly enriched.

However, when including the 2kb flanking regions of HERV S2 and performing the GO enrichment on the 5518 identified genes, the six GO terms shown in table \ref{tab:hervS2.2kb.enrichment} are significantly overrepresented. They are all connected to defense response, especially against viruses.

\begin{table}[h!]
  \begin{center}
  \resizebox{\textwidth}{!}{
    \pgfplotstabletypeset[
    col sep=tab, 
    header=true,    
	columns={Term ID, Term, p, fdr},
	columns/Term ID/.style={column type=l,string type},
	columns/Term/.style={column type=l,string type},
	every head row/.style={before row=\hline, after row=\hline},
	every last row/.style={after row=\hline}]
    {../tables/hervS2.2kb.expr.enrichment.tsv}}
  \end{center}        
	\caption{Significantly enriched GO biological process terms among genes overlapping with HERV S2.}
	\label{tab:expr.overlap.counts}
\end{table}
%Wrong hervS2.2kb


The results for all HERV sets and including flanking regions is shown in table \ref{tab:expr.overlap.counts}.

\begin{table}[h!]
  \begin{center}
    \pgfplotstabletypeset[
    col sep=tab, 
    header=true,    
	columns={Set,S1,S1.1kb,S1.2kb,S2,S2.1kb,S2.2kb,S3,S3.1kb,S3.2kb},
	columns/Set/.style={column type=l,string type},
	every head row/.style={before row=\hline, after row=\hline},
	every last row/.style={after row=\hline}]
    {../tables/expr.overlap.overview.tsv}
  \end{center}        
	\caption{Number of expression probes overlapping with different HERV sets and flanking regions. "Pairs" is the total number of overlaps occurring, "HERVs" is the number of distinct HERV elements that have an overlap with any of the expression probes, Probes describes the number of distinct expression probes that overlap with the HERV elements or their flanking regions, "Genes" is the number of distinct Genes that are annotated to these probes.}
	\label{tab:expr.overlap.counts}
\end{table}

\subsubsection{Methylation}
17077 CpG sites, equaling 3.52\% of all measured CpGs, were found within HERV S2 and 12871 distinct HERV elements (2.10\%) contain at least one interogated CpG site.


%The variance of the measurements of these sites, which is shown in figure \ref{fig:}, is lower on average.

The number of CpG sites within all HERV sets and including flanking regions is shown in table \ref{tab:meth.overlap.counts}

\begin{table}[h!]
  \begin{center}
    \resizebox{\textwidth}{!}{
    \pgfplotstabletypeset[
    col sep=tab, 
    header=true,    
	columns={Set,S1,S1.1kb,S1.2kb,S2,S2.1kb,S2.2kb,S3,S3.1kb,S3.2kb},
	columns/Set/.style={column type=l,string type},
	every head row/.style={before row=\hline, after row=\hline},
	every last row/.style={after row=\hline}]
    {../tables/meth.overlap.overview.tsv}}
  \end{center}        
	\caption{Number of CpGs overlapping with different HERV sets and flanking regions. "Pairs" is the total number of overlaps occurring, "HERVs" is the number of distinct HERV elements that have an overlap with any of the expression probes, "CpGs" is the number of distinct CpG sites that lie within the HERV elements or their flanking regions.}
	\label{tab:meth.overlap.counts}
\end{table} 
\subsubsection{Genotypes}
A total of 890780 the considered SNPs are located within elements of HERV S2. This constitutes 9.34\% of all SNPs. These SNPs are found in 330744 distinct HERV elements. Therefore, 53.99\% contain at least one SNP.

The results for all sets and flanking regions are shown in table \ref{tab:snp.overlap.counts}

\begin{table}[h!]
  \begin{center}
    \resizebox{\textwidth}{!}{
    \pgfplotstabletypeset[
    col sep=tab, 
    header=true,    
	columns={Set,S1,S1.1kb,S1.2kb,S2,S2.1kb,S2.2kb,S3,S3.1kb,S3.2kb},
	columns/Set/.style={column type=l,string type},
	every head row/.style={before row=\hline, after row=\hline},
	every last row/.style={after row=\hline}]
    {../tables/snp.overlap.overview.tsv}}
  \end{center}        
	\caption{Number of SNPs overlapping with different HERV sets and flanking regions. "Pairs" is the total number of overlaps occurring, "HERVs" is the number of distinct HERV elements that have an overlap with any of the expression probes, "SNPs" is the number of distinct considered SNPs that lie within the HERV elements or their flanking regions.}
	\label{tab:snp.overlap.counts}
\end{table} 

\subsubsection{Chromatin states}
Data in /storage/groups/groups\textunderscore epigenreg/users/julian.schmidt ...

\subsection{eQTLs}
Associations were calculated for 156.4 millions cis acting SNP-expression probe pairs and 456.1 billion possible pairs in trans. 

812147 of cis-pairs were significantly associated with p-values of 1e-6 or less. These significant eQTLs have a false discovery rate of less than 1.93e-4. A total of 551728 distinct SNPs and 4903 expression probes were part of at least one cis-eQTL. 4145 of these probes are annotated to 3552 different genes. The remaining 
758 can not be assigned to a specific gene.

There were a total of 1511235 significant trans-eQTL with an p-value of less than 1e-8 and a FDR of less than 3.02e-4. These are made up by 229332 distinct SNPs and 21338 expression probes. 11505 of the probes found in at least one trans-eQTL are annotated to 9389 different genes, while 9767 probes are not assigned to a gene.

In a previous work Schramm et al. \cite{} calculated eQTLs on expression and genotype data from 890 KORA F4 samples. They found significant cis-eQTLs for 4116 probes and considered only the pair of the most strongly associated SNPs with the probe. Of these pairs 2680 were also found in my analysis. 

Figure \ref{fig:global.eqtl.heatmap}A shows the fraction of significant to all possible  SNP-expression probe pairs, mapped to 10 Mpb windows. As expected the values at the diagonal are comparatively higher, containing all cis-eQTLs. 

The signatures of wildtype, heterogenous and homogenous mutated SNP site for the two eQTLs with the best and worst p-values in cis and trans are shown in figure \ref{fig:best.worst.eqtl.boxplot}. All eQTLs show clear differences in expression between different genotypic variants. 

45862 of the SNPs and 166 of the expression probes found in at least one cis-eQTL are located within HERV S2. When considering all significant associations, where either the SNP or the expression probe lie within a HERV element, a total of 112604 pairs remain. There are 5855 cis-eQTLs constituting one of 4748 SNPs within a HERV element controlling one of 128 expression probes, that overlap with a HERV element. 

When limiting trans-eQTLs to the ones that contain a HERV S2 SNP (23396) and/or expression probe (1227), 240301 remain. In 9177 of these pairs both are HERV related. 

The fraction of HERV S2 related SNP-expression probe pairs that proved to be significantly associated, respective to their genomic locations is shown in figure \ref{fig:global.eqtl.heatmap}B. 

A GO enrichment was performed on all 8211 genes associated to HERV related eQTLs using the set of all genes found in eQTLs as background. However, there were no significantly enriched genes. 

\subsection{eQTMs}
When calculating expression quantitative trait methylation there were a total of 13.88 millions CpG-expression probe pairs within a distance of 50 Kpb or less of each other. The number of potential trans-acting pairs equaled around 23.22 billions. 

Calculating eQTMs with a significance threshold of 10e-6 for cis resulted in 8187 significant associations (fdr<1.7e-3) consisting of 5957 distinct CpG sites and 1959 different expression probes. 1658 of these probes are annotated to 1461 genes, while there are no gene annotations for the remainder. 

361485 of the potential trans-action CpG-expression probe pairs proved to be significant at a p-value threshold of 10e-8 (fdr<6.5e-4). These trans-eQTLs are made up by 48206 CpGs and 11673 expression probes, for which 5738 gene annotations are available. 

The fractions of potential CpG-expression probe pairs that were significantly associated between their genomic locations are shown in figure \ref{fig:global.eqtm.heatmap}A. 

HERV S2 contains 311 CpG sites and 420 expression probes, that were present in cis-eQTMs. A total of 738 cis-eQTMs are related to the set. 33 of these are associations between one of 33 CpG sites within a HERV element and one of 14 expression probes. 

Considering trans-eQTMs, 1109 significantly associated CpGs lie within HERV S2 and 5422 expression probes overlapping a HERV element are part of a trans-eQTM.

The results of GO biological process enrichment performed on the xxxx genes that the expression probes found in HERV S2 related eQTMs are shown in table \ref{fig:hervS2.eqtm.enrichment}. 
%whatever is in that fricking table


\subsection{meQTLs}

\subsection{HERV related regulatory networks}
\subsubsection{Data collection}


\newpage
\section{Discussion}

\newpage
\section{Conclusion and Outlook}

\newpage
\bibliography{mybib}{}
\bibliographystyle{unsrt}

\end{document}